### PR TITLE
gpio: Python: change gpioChip to match a generic /dev/gpiochip* number

### PR DESCRIPTION
**Pull request by AllanKamimura**
changed the sample example to include devices with GPIO in any gpio bank, not only **gpiochip0**
I tested on apalis-imx8  with TorizonCore 6.6.1+build.14 (kirkstone) and the GPIO are in **gpiochip2**

-  docker-compose.yml
  -  expose all gpiochip devices
  
- main.py
  - when gpioChip == None, added an outer loop to iterate over all /dev/gpiochip*, before the inner loop to iterate over all line_offset
  - added string format to make the **print** output similar to the **gpioinfo** output

### CODE REVIEW

File: gpio/python/gpioPython/src/main.py [hunk 1]
```diff
@@ -2,38 +2,39 @@
...
+                if line.is_used():
+                    line_consumer = line.consumer()
+                else: 
+                    line_consumer = "unused"
+
+                if line.direction() == "1":
+                    line_direction = "input"
+                else:  
+                    line_direction = "output"
+                print(f"Line {line_offset:<2}:  {line_name:<15}  {line_consumer:<20}  {line_direction:<10}")
```
Comment (andreriesco) [inline gpio/python/gpioPython/src/main.py:32]: What these `<` do?

File: gpio/python/gpioPython/docker-compose.yml [hunk 2]
```diff
@@ -8,13 +8,21 @@ services:
     ports:
       - 6502:6502
       - 6512:6512
-    devices:
-      - "/dev/gpiochip0"
-    
+    device_cgroup_rules:
+      - "c 254:* rmw"
+    volumes:
+      - type: bind
```
Comment (andreriesco) [inline gpio/python/gpioPython/docker-compose.yml:14]: This one is a trickier one. The idea on the `docker-compose.yml` is to follow good practices and give access just to the directories on the host needed for the GPIOs to work, so it is not good to give access to all files in the `dev` directory.

File: gpio/python/gpioPython/src/main.py (hunk 1 above)
Comment (AllanKamimura) [reply gpio/python/gpioPython/src/main.py:32]: it's for string formatting
`< : `left align
`> : `right align
`^ : `center align

File: gpio/python/gpioPython/docker-compose.yml (hunk 2 above)
Comment (otaviocp) [reply gpio/python/gpioPython/docker-compose.yml:14]: I think in that case, there is no problem mounting the `/dev` directory. The container will not be able to access devices listed there, except for those explicitly allowed through the 'device_cgroup_rules' property ([Runtime privilege and Linux capabilities](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities)).

In our case, we only allowed devices with major number 254, which corresponds to the gpiochips.

Comment (andreriesco) [reply gpio/python/gpioPython/docker-compose.yml:14]: So only the `/dev/gpiochip*` will be visible/available? 
```
torizon@verdin-imx8mp-14773124:~$ ls /dev
autofs           emmc-boot1  i2c-1         loop4               null   ptyp9  ram2    spidev1.0  tty18  tty31  tty45  tty59    ttymxc2  ttype     vcsa2        verdin-adc4           video0
block            emmc-part1  i2c-2         loop5               port   ptypa  ram3    stderr     tty19  tty32  tty46  tty6     ttyp0    ttypf     vcsa3        verdin-i2c-on-module  video1
btrfs-control    fb0         i2c-3         loop6               ppp    ptypb  ram4    stdin      tty2   tty33  tty47  tty60    ttyp1    ubi_ctrl  vcsa4        verdin-i2c1           vsi_daemon_ctrl
bus              fd          i2c-6         loop7               ptmx   ptypc  ram5    stdout     tty20  tty34  tty48  tty61    ttyp2    uinput    vcsa5        verdin-i2c2           watchdog
caam-keygen      full        iio:device0   mapper              ptp0   ptypd  ram6    tty        tty21  tty35  tty49  tty62    ttyp3    urandom   vcsa6        verdin-i2c3           watchdog0
cec0             fuse        initctl       mem                 pts    ptype  ram7    tty0       tty22  tty36  tty5   tty63    ttyp4    v4l       vcsu         verdin-i2c4           zero
char             galcore     input         mmcblk2             ptyp0  ptypf  ram8    tty1       tty23  tty37  tty50  tty7     ttyp5    vcs       vcsu1        verdin-spi-cs0        zram0
console          gpiochip0   kmsg          mmcblk2boot0        ptyp1  ram0   ram9    tty10      tty24  tty38  tty51  tty8     ttyp6    vcs1      vcsu2        verdin-uart1
cpu_dma_latency  gpiochip1   kvm           mmcblk2boot1        ptyp2  ram1   random  tty11      tty25  tty39  tty52  tty9     ttyp7    vcs2      vcsu3        verdin-uart2
cuse             gpiochip2   log           mmcblk2p1           ptyp3  ram10  rfkill  tty12      tty26  tty4   tty53  ttyS0    ttyp8    vcs3      vcsu4        verdin-uart3
disk             gpiochip3   loop-control  mmcblk2rpmb         ptyp4  ram11  rtc     tty13      tty27  tty40  tty54  ttyS1    ttyp9    vcs4      vcsu5        verdin-watchdog
dma_heap         gpiochip4   loop0         mqueue              ptyp5  ram12  rtc0    tty14      tty28  tty41  tty55  ttyS2    ttypa    vcs5      vcsu6        verdin-watchdog-soc
dri              hugepages   loop1         mxc_hantro          ptyp6  ram13  rtc1    tty15      tty29  tty42  tty56  ttyS3    ttypb    vcs6      verdin-adc1  vfio
emmc             hwrng       loop2         mxc_hantro_vc8000e  ptyp7  ram14  shm     tty16      tty3   tty43  tty57  ttymxc0  ttypc    vcsa      verdin-adc2  vga_arbiter
emmc-boot0       i2c-0       loop3         net                 ptyp8  ram15  snd     tty17      tty30  tty44  tty58  ttymxc1  ttypd    vcsa1     verdin-adc3  vhci
```

Comment (andreriesco) [reply gpio/python/gpioPython/docker-compose.yml:14]: Also, these are the same in practice?
```
    device_cgroup_rules:
      - "c 254:* rmw"
    volumes:
      - type: bind
        source: /dev/gpiochip0
        target: /dev/gpiochip0
```
```
 devices:
      - "/dev/gpiochip0"
```

Comment (otaviocp) [reply gpio/python/gpioPython/docker-compose.yml:14]: Yes, those examples give the same result. It's also possible to enable access only to the gpiochip0 using its minor number:
```
  device_cgroup_rules:
    - "c 254:1 rmw"
  volumes:
    - type: bind
      source: /dev/
      target: /dev/
```

Here, we would like to mount `/dev` and allow access to all gpiochips to ensure the sample works on any module.
Running `ls -l /dev | grep gpio` on the Verdin AM62:
```
crw-rw-r-- 1 root gpio    254,   0 Jun 17 18:07 gpiochip0
crw-rw-r-- 1 root gpio    254,   1 Jun 17 18:07 gpiochip1
crw-rw-r-- 1 root gpio    254,   2 Jun 17 18:07 gpiochip2

```
Running `ls -l /dev | grep gpio` on the Apalis iMX8:
```
crw-rw-r-- 1 root gpio    254,   0 Jun 17 18:09 gpiochip0
crw-rw-r-- 1 root gpio    254,   1 Jun 17 18:09 gpiochip1
crw-rw-r-- 1 root gpio    254,   2 Jun 17 18:09 gpiochip2
crw-rw-r-- 1 root gpio    254,   3 Jun 17 18:09 gpiochip3
crw-rw-r-- 1 root gpio    254,   4 Jun 17 18:09 gpiochip4
crw-rw-r-- 1 root gpio    254,   5 Jun 17 18:09 gpiochip5
crw-rw-r-- 1 root gpio    254,   6 Jun 17 18:09 gpiochip6
crw-rw-r-- 1 root gpio    254,   7 Jun 17 18:09 gpiochip7
crw-rw-r-- 1 root gpio    254,   8 Jun 17 18:09 gpiochip8
crw-rw-r-- 1 root gpio    254,   9 Jun 17 18:09 gpiochip9

```
Other devices are visible, but cannot be accessed.

Comment (andreriesco) [reply gpio/python/gpioPython/docker-compose.yml:14]: Sounds good the change, the only thing would be to add on the [README](https://github.com/toradex/torizon-samples/blob/bookworm-new/gpio/python/gpioPython/README.md) the info that on production the person should expose just the gpiochip the application needs, and how to do this, by adding the `/dev/gpiochip` on the bind-mount or by adding the number of the gpiochip on the  cgroup_rules `- "c 254:1 rmw"`

Comment (AllanKamimura) [reply gpio/python/gpioPython/docker-compose.yml:14]: ok, so the conclusion is

we keep this part, to make this sample works for any module
```
    device_cgroup_rules:
      - "c 254:* rmw"
    volumes:
      - type: bind
        source: /dev
        target: /dev
```

and include some steps instructions for the user in the README

1. For evaluation, run the example and find the chipset GPIO bank used by the GPIO pins the user is going to use
2. Modify the `/src/main.py` with the `gpioChip` and `gpiolineName`
3. And add a info section saying that "for production, it's a best practices to expose only the devices the user is going to use"

**Comment by AllanKamimura** [comment]
this gpio python package was updated, this one seems to be the official package now
https://pypi.org/project/gpiod/

I updated the main script for this new package

**Comment by andreriesco** [comment]
@AllanKamimura I believe there are some files missing in the `try-update-template`, like the `tasks.json`.

**Comment by AllanKamimura** [comment]
Hi @andreriesco 
I tried to re-re-run the `try-update-template`, I think it is working now

**Comment by AllanKamimura** [comment]
closing this MR in favor of https://github.com/toradex/torizon-samples/pull/29

**Pull request closed without merging**